### PR TITLE
feat(meta): split stateful operators into different fragments

### PR DESCRIPTION
**Pull request by skyzh**
Signed-off-by: Alex Chi <iskyzh@gmail.com>

## What's changed and what's your intention?

This PR adds a rewriter on meta service to split stateful operators into different fragments.

before:

image

now:

image

changes:

* all stream plan nodes now have a schema. otherwise meta node cannot insert exchange node with the correct exchange fields.
* remove a test suite. e2e is already good enough to test things out, no need to manually construct all plan nodes.
* remove `fields` (schema) from exchange node.
* Rust frontend will now automatically add a schema for all plan nodes. Java frontend won't, but as it will be deprecated soon, we won't refactor it.
* a new exchange type called `no_shuffle`, that is used between such splits. No-shuffle exchange will add 1-to-1 channel between upstream actors and downstream actors within a fragment.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)

close https://github.com/singularity-data/risingwave/issues/1745

### CODE REVIEW

**Comment by skyzh** [comment]
One might ask why I don't do this on frontend. That's because the fragmenter is on meta, so such rewrites can only be efficiently done on meta. Also, for constructing delta join plans, such rewrites are also useful.

**Comment by fuyufjh** [comment]
> This PR adds a rewriter on meta service to split stateful operators into different fragments.

1. Why need to rewrite the plan? i.e. adding another `Exchange`
2. Why not do this in frontend?

File: src/meta/src/stream/fragmenter.rs [hunk 1]
```diff
@@ -141,9 +147,64 @@ where
...
+                            column_indices: vec![],
+                        };
+
+                        StreamNode {
+                            pk_indices: child_node.pk_indices.clone(),
+                            fields: child_node.fields.clone(),
+                            node: Some(Node::ExchangeNode(ExchangeNode {
+                                strategy: Some(strategy.clone()),
+                            })),
+                            operator_id: 10000000 + child_node.operator_id,
```
Comment (yezizp2012) [inline src/meta/src/stream/fragmenter.rs:185]: 😂

File: src/meta/src/stream/fragmenter.rs [hunk 2]
```diff
@@ -89,8 +91,12 @@ where
         stream_node: &StreamNode,
         ctx: &mut CreateMaterializedViewContext,
     ) -> Result<BTreeMap<FragmentId, Fragment>> {
+        let stream_node = stream_node.clone();
+
         // Generate fragment graph and seal
         self.generate_fragment_graph(stream_node)?;
+        // The stream node might be rewritten after this point. Don't use `stream_node` anymore.
```
Comment (MrCroxx) [inline src/meta/src/stream/fragmenter.rs:98]: Can we simply take its ownership in `generate_graph`? Does the caller still needs the origin plan anymore? 🤔

Comment (skyzh) [reply src/meta/src/stream/fragmenter.rs:98]: Yes. It won't need. After this step, the plan is now fragments in fragment graph.

File: src/meta/src/stream/fragmenter.rs (hunk 1 above)
Comment (skyzh) [reply src/meta/src/stream/fragmenter.rs:185]: This id is not important -- it won't be stored in meta or used by something else, and it's not even globally unique. So it's okay to assign such value here.

**Review by yezizp2012 — APPROVED** [review]

File: src/meta/src/stream/fragmenter.rs [hunk 3]
```diff
@@ -141,9 +147,64 @@ where
             .collect::<Result<BTreeMap<_, _>>>()
     }
 
+    /// Do some dirty rewrites on meta. Currently, it will split stateful operators into two
+    /// fragments.
+    fn rewrite_stream_node(&self, stream_node: StreamNode) -> Result<StreamNode> {
+        self.rewrite_stream_node_inner(stream_node, false)
+    }
+
+    fn rewrite_stream_node_inner(
```
Comment (yezizp2012) [inline src/meta/src/stream/fragmenter.rs:156]: Can `rewrite_stream_node_inner` be done in frontend? It might be more reasonable. 🤔

Comment (skyzh) [reply src/meta/src/stream/fragmenter.rs:156]: Then we will have to move the whole fragmenter to frontend, and having a proto to pass fragment among nodes 🤣 Currently, fragments are only intermediate representation that are generated on meta node.

**Pull request merged**